src/builtin_path.cpp

```diff
@@ -0,0 +1,777 @@
+    options_t opts;
+    opts.type_valid = true;
+    opts.perm_valid = true;
+    opts.invert_valid = true;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        if ((!opts.invert || (!opts.have_perm && !opts.have_type)) && filter_path(opts, *arg)) {
+            // If we don't have filters, check if it exists.
+            // (for match this is done by the glob already)
+            if (!opts.have_type && !opts.have_perm) {
+                bool ok = !waccess(*arg, F_OK);
+                if (ok == opts.invert) continue;
+            }
+
+            path_out(streams, opts, *arg);
+            n_transformed++;
+            if (opts.quiet) return STATUS_CMD_OK;
+        }
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+// Keep sorted alphabetically
+static constexpr const struct path_subcommand {
+    const wchar_t *name;
+    int (*handler)(parser_t &, io_streams_t &, int argc,  //!OCLINT(unused param)
+                   const wchar_t **argv);                 //!OCLINT(unused param)
+} path_subcommands[] = {
+    // TODO: Which operations do we want?
+    // TODO: "base" or "basename"?
+    {L"base", &path_base},
+    {L"dir", &path_dir},
+    {L"expand", &path_expand},
+    {L"extension", &path_extension},
+    {L"filter", &path_filter},
+    {L"match", &path_match},
+    {L"normalize", &path_normalize},
+    {L"real", &path_real},
+    {L"strip-extension", &path_strip_extension},
+};
+ASSERT_SORTED_BY_NAME(path_subcommands);
+
+/// The string builtin, for manipulating strings.
+maybe_t<int> builtin_path(parser_t &parser, io_streams_t &streams, const wchar_t **argv) {
+    const wchar_t *cmd = argv[0];
+    int argc = builtin_count_args(argv);
+    if (argc <= 1) {
+        streams.err.append_format(BUILTIN_ERR_MISSING_SUBCMD, cmd);
+        builtin_print_error_trailer(parser, streams.err, L"path");
+        return STATUS_INVALID_ARGS;
+    }
+
+    if (std::wcscmp(argv[1], L"-h") == 0 || std::wcscmp(argv[1], L"--help") == 0) {
+        builtin_print_help(parser, streams, L"path");
+        return STATUS_CMD_OK;
+    }
+
+    const wchar_t *subcmd_name = argv[1];
+    const auto *subcmd = get_by_sorted_name(subcmd_name, path_subcommands);
+    if (!subcmd) {
+        streams.err.append_format(BUILTIN_ERR_INVALID_SUBCMD, cmd, subcmd_name);
+        builtin_print_error_trailer(parser, streams.err, L"path");
+        return STATUS_INVALID_ARGS;
+    }
+
+    if (argc >= 3 && (std::wcscmp(argv[2], L"-h") == 0 || std::wcscmp(argv[2], L"--help") == 0)) {
+        wcstring path_dash_subcommand = wcstring(argv[0]) + L"-" + subcmd_name;
+        builtin_print_help(parser, streams, path_dash_subcommand.c_str());
+        return STATUS_CMD_OK;
+    }
```

Comment:
Quite unfortunate that we need to keep the option in case we receive `"some\npath"`.

Actually, wouldn't it be better to drop the null-in and null-out feature?
I think it's rarely needed, I don't think users have asked for it?

- We probably have many scripts that break on paths with newlines.
  I doubt we want to change them for such an unlikely case.
- Keep in mind that in other shells `find -print0 | xargs -0 rm` is probably most often used in practice to protect against spaces in filenames. We don't have that problem.
- In case someone needs it, they can just pass it as argument, like
  `for filename_with_newlines in (find -print0 | string split0 -- $data); path extension -- $filename_with_newlines`; end.
  - This has the advantage that it works for any command.

The advantage is that the documentation becomes smaller.  I believe most
users won't ever need these options, so it's better if they don't have to
think about them.

Sorry that this suggestion comes so late, I wasn't really active in the planning stage.

Reply:
>Quite unfortunate that we need to keep the option in case we receive "some\npath".

>Actually, wouldn't it be better to drop the null-in and null-out feature?
I think it's rarely needed, I don't think users have asked for it?

Yes, we do. Both for *actual, theoretical completeness* (but yes, just abolish newlines in paths and everything becomes about ~300% easier! if posix could simply do that that would be great!), and for integration with things meant for filenames.

And because this is *already included* and *not a lot of code*, we can simply keep it in. I do not believe the three or so sentences in the docs are a problem.